core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,425 @@
+      outBuffer.clear()
+      decryptor.decrypt(inBuffer, outBuffer)
+      inBuffer.clear()
+      outBuffer.flip()
+      if (padding > 0) {
+        outBuffer.position(padding)
+      }
+    }
+  }
+
+  /**
+   * This method is executed immediately after decryption. Check whether
+   * decryptor should be updated and recalculate padding if needed.
+   */
+  def afterDecryption(decryptor: Decryptor, inBuffer: ByteBuffer, position: Long,
+                      iv: Array[Byte]): Byte = {
+    var padding: Byte = 0
+    if (decryptor.isContextReset) {
+      updateDecryptor(decryptor, position, iv)
+      padding = getPadding(position)
+      inBuffer.position(padding)
+    }
+    padding
+  }
+
+  def getCounter(position: Long): Long = {
+    position / codec.getCipherSuite.algoBlockSize
+  }
+
+  def getPadding(position: Long): Byte = {
+    (position % codec.getCipherSuite.algoBlockSize).asInstanceOf[Byte]
+  }
+
+  /** Calculate the counter and iv, update the decryptor. */
+  def updateDecryptor(decryptor: Decryptor, position: Long, iv: Array[Byte]) {
+    val counter = getCounter(position)
+    codec.calculateIV(initIV, counter, iv)
+    decryptor.init(key, iv)
+  }
+
+  /**
+   * Reset the underlying stream offset clear [[inBuffer]] and
+   * [[outBuffer]]. This Typically happens during seek operation
+   * or [[skip]].
+   */
+  def resetStreamOffset(offset: Long) {
+    streamOffset = offset
+    inBuffer.clear()
+    outBuffer.clear()
+    outBuffer.limit(0)
+    updateDecryptor(decryptor, offset, iv)
+    padding = getPadding(offset)
+    inBuffer.position(padding)
+  }
+
+  override def close: Unit = {
+    if (!closed) {
+      super.close()
+      freeBuffers()
+      closed = true
+    }
+  }
+
+  /** Skip n bytes */
+  override def skip(nVal: Long): Long = {
+    var n = nVal
+    Preconditions.checkArgument(n >= 0, "Negative skip length.", new Array[String](1))
+    checkStream()
+    if (n == 0) {
+      0
+    } else if (n <= outBuffer.remaining()) {
+      val pos = outBuffer.position() + n
+      outBuffer.position(pos.toInt)
+      n
+    } else {
+      n -= outBuffer.remaining()
+      var skipped: Long = in.skip(n)
+      if (skipped < 0) {
+        skipped = 0
+      }
+      val pos: Long = streamOffset + skipped
+      skipped += outBuffer.remaining()
+      resetStreamOffset(pos)
+      skipped
+    }
+  }
+
+  /** ByteBuffer read. */
+  override def read(buf: ByteBuffer): Int = {
+    checkStream()
+    if (isReadableByteChannel) {
+      val unread = outBuffer.remaining()
+      if (unread > 0) {
+        val toRead = buf.remaining()
+        if (toRead <= unread) {
+          val limit = outBuffer.limit()
+          outBuffer.limit(outBuffer.position() + toRead)
+          buf.put(outBuffer)
+          outBuffer.limit(limit)
+          return toRead
+        } else {
+          buf.put(outBuffer)
+        }
+      }
+      val pos = buf.position()
+      val count = (in.asInstanceOf[ReadableByteChannel]).read(buf)
```

Comment:
In the `read` method above, you have all this code to detect whether this call will throw a `UnsupportedOperationException`. Why is that not necessary here?